interfaces/apparmor/spec.go

```diff
@@ -92,7 +92,19 @@ func (spec *Specification) AddSnippet(snippet string) {
 
 // AddUpdateNS adds a new apparmor snippet for the snap-update-ns program.
 func (spec *Specification) AddUpdateNS(snippet string) {
-	spec.updateNS = append(spec.updateNS, snippet)
+	spec.updateNS.Put(snippet)
+}
+
+// EmitUpdateNSFunc returns a function for emitting update-ns snippets.
+func (spec *Specification) EmitUpdateNSFunc() func(f string, args ...interface{}) {
```

Comment:
Following my comment to the earlier PR regarding the indirection here, cannot this simply become:
```
diff --git a/interfaces/apparmor/spec.go b/interfaces/apparmor/spec.go
index 93f5429c69..2c8f933acb 100644
--- a/interfaces/apparmor/spec.go
+++ b/interfaces/apparmor/spec.go
@@ -96,10 +96,8 @@ func (spec *Specification) AddUpdateNS(snippet string) {
 }

 // EmitUpdateNSFunc returns a function for emitting update-ns snippets.
-func (spec *Specification) EmitUpdateNSFunc() func(f string, args ...interface{}) {
-       return func(f string, args ...interface{}) {
-               spec.AddUpdateNS(fmt.Sprintf(f, args...))
-       }
+func (spec *Specification) EmitUpdateNSFunc(f string, args ...interface{}) {
+       spec.AddUpdateNS(fmt.Sprintf(f, args...))
 }

 // UpdateNSIndexOf returns the index of a previously added snippet.
@@ -158,7 +156,7 @@ func (spec *Specification) AddLayout(si *snap.Info) {
                sort.Strings(spec.snippets[tag])
        }

-       emit := spec.EmitUpdateNSFunc()
+       emit := spec.EmitUpdateNSFunc

        // Append update-ns snippets that allow constructing the layout.
        for _, path := range paths {
diff --git a/interfaces/builtin/appstream_metadata.go b/interfaces/builtin/appstream_metadata.go
index ffc2711287..4a235c45b2 100644
--- a/interfaces/builtin/appstream_metadata.go
+++ b/interfaces/builtin/appstream_metadata.go
@@ -75,7 +75,7 @@ func (iface *appstreamMetadataInterface) AppArmorConnectedPlug(spec *apparmor.Sp
        spec.AddSnippet(appstreamMetadataConnectedPlugAppArmor)

        // Generate rules to allow snap-update-ns to do its thing
-       emit := spec.EmitUpdateNSFunc()
+       emit := spec.EmitUpdateNSFunc
        for _, target := range appstreamMetadataDirs {
                source := "/var/lib/snapd/hostfs" + target
                emit("  # Read-only access to %s\n", target)
diff --git a/interfaces/builtin/content.go b/interfaces/builtin/content.go
index b15df7faa9..e8bd912e3e 100644
--- a/interfaces/builtin/content.go
+++ b/interfaces/builtin/content.go
@@ -214,7 +214,7 @@ func mountEntry(plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot,
 func (iface *contentInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
        contentSnippet := bytes.NewBuffer(nil)
        writePaths := iface.path(slot, "write")
-       emit := spec.EmitUpdateNSFunc()
+       emit := spec.EmitUpdateNSFunc
        if len(writePaths) > 0 {
                fmt.Fprintf(contentSnippet, `
 # In addition to the bind mount, add any AppArmor rules so that
diff --git a/interfaces/builtin/desktop.go b/interfaces/builtin/desktop.go
index 7bb8c9c89c..944de43ca1 100644
--- a/interfaces/builtin/desktop.go
+++ b/interfaces/builtin/desktop.go
@@ -231,7 +231,7 @@ func (iface *desktopInterface) AppArmorConnectedPlug(spec *apparmor.Specificatio
        spec.AddSnippet(desktopConnectedPlugAppArmor)

        // Allow mounting document portal
-       emit := spec.EmitUpdateNSFunc()
+       emit := spec.EmitUpdateNSFunc
        emit("  # Mount the document portal\n")
        emit("  mount options=(bind) /run/user/[0-9]*/doc/by-app/snap.%s/ -> /run/user/[0-9]*/doc/,\n", plug.Snap().InstanceName())
        emit("  umount /run/user/[0-9]*/doc/,\n\n")
```

(kept emit := ... in place to reduce the diff and not touch all the emit(..) lines).

The tests seem to be happy.

Reply:
Yeah, it can _now_ :-)

As this code was shaping, doing what you asked for would introduce observable differences and de-rail the argument of mechanical change that can be easier to review.

We can streamline this in this PR or in the follow-up.

Reply:
Allright. And as mentioned earlier, the Gen* helpers could become methods of spec and drop _emit_ func argument), .e.g.
```
spec.GenWritableProfile(target, 3)
```

Reply:
I wouldn't do that because they generate apparmor rules that could equally well be placed in `snap-update-ns.pkg` as in `snap.pkg.app`. We could make it so that the emit function is simply AddUpdateNS so there's less glue code though.